cf/test/test_Data.py

```diff
@@ -131,6 +131,293 @@ class DataTest(unittest.TestCase):
     #    test_only = ['test_Data_clip']
     #    test_only = ['test_Data__init__dtype_mask']
 
+    def test_Data_equals(self):
+        if self.test_only and inspect.stack()[0][3] not in self.test_only:
+            return
+
+        shape = 3, 4
+        a = np.arange(12).reshape(*shape)
+
+        d = cf.Data(a, "m")
+        self.assertTrue(d.equals(d))  # also do self-equality checks!
+
+        d2 = cf.Data(a.astype(np.float32), "m")  # different datatype to d
+        self.assertTrue(d2.equals(d2))
+        with self.assertLogs(level=cf.log_level().value) as catch:
+            self.assertFalse(d2.equals(d))
+            self.assertTrue(
+                any(
+                    "Data: Different data types: float32 != int64" in log_msg
+                    for log_msg in catch.output
+                )
+            )
+
+        e = cf.Data(a, "s")  # different units to d
+        self.assertTrue(e.equals(e))
+        with self.assertLogs(level=cf.log_level().value) as catch:
+            self.assertFalse(e.equals(d))
+            self.assertTrue(
+                any(
+                    "Data: Different Units (<Units: s>, <Units: m>)" in log_msg
+                    for log_msg in catch.output
+                )
+            )
+
+        f = cf.Data(np.arange(12), "m")  # different shape to d
+        self.assertTrue(f.equals(f))
+        with self.assertLogs(level=cf.log_level().value) as catch:
+            self.assertFalse(f.equals(d))
+            self.assertTrue(
+                any(
+                    "Data: Different shapes: (12,) != (3, 4)" in log_msg
+                    for log_msg in catch.output
+                )
+            )
+
+        g = cf.Data(np.ones(shape, dtype="int64"), "m")  # different values
+        self.assertTrue(g.equals(g))
+        with self.assertLogs(level=cf.log_level().value) as catch:
+            self.assertFalse(g.equals(d))
+            self.assertTrue(
+                any(
+                    "Data: Different array values" in log_msg
+                    for log_msg in catch.output
+                )
+            )
+
+        # Test NaN and inf values
+        d3 = cf.Data(a.astype(np.float64), "m")
+        h = cf.Data(np.full(shape, np.nan), "m")
+        # TODODASK: is this OK given that NaN in NumPy aren't equal to e/o
+        # or should this be assertTrue to expect equality behaviour?
+        self.assertFalse(h.equals(h))
```

Comment:
#### Outstanding issue 2

Notice that this is the lone self-equality which is `assertFalse` rather than `assertTrue`, and (as with everything except those assertions discussed in my other in-line comment for issue 1) it passes at the moment, because it is an array with NaN elements and naturally in NumPy NaN is not equal to itself.

Any cf.Data object not being equal to itself seems either wrong or potentially confusing to me. Do you think we should keep this behaviour, or change it so that NaNs are considered equal (eay enough to do with the `equal_nan` kwarg to `allclose`, etc.)?

I haven't converged on a strong preference either way, though I do think if we keep the behaviour of the self-equality being False we should at least provide a warning, e.g. via the logging: something like "The data array contains NaN elements which are not considered equal to themselves". And, of course, there's the option of providing our own `equal_nan` kwarg for flexibility for users to choose?

Reply:
I'll need to think about this, but will do after submitting the other review comments

Reply:
It was mentioned elsewhere that "`cf.Data.equals` has, up till now at least, short circuited if `self is other`". But this is not the case_, i.e. `cf.Data,equals` always compares the data values, even when "self is other"

It is the case, however, that `cfdm.Data.equals` short circuits when self is other, and I incorrectly  assumed that this was inherited by cf. I think that I probably meant to make sure that `cf.Data,equals` does short circuit in this case, but overlooked it.

So in the status quo:
```python
>>> import numpy as np; import cf, cfdm
>>> h = cfdm.Data(np.full((3, 4), np.nan), "m")    ## cfdm
>>> h.equals(h, verbose=-1)
True
>>> h = cf.Data(np.full((3, 4), np.nan), "m")      ## cf
>>> h.equals(h, verbose=-1)
Data: Different array values (atol=2.220446049250313e-16, rtol=2.220446049250313e-16)
False
```

I.e. cfdm short cricuits and gives `True`, but cf does not and give `False`, because numpy itself deems NaNs no equal.

As this falls in the realms of cock up rather than conspiracy, I think that we should consider  ourselves free to define a consistent behaviour across cfdm and cf, regardless of whcih each library does at present.

That's a good place to stop. Now I'm up to speed I can start to think about the problem as stated ...

Reply:
Agreed. Whatever we go for should be the case for both libraries. Good thinking to consider the cfdm behaviour whilst we are considering this, too.